File: app/src/main/java/com/salman/bitclock/ui/stopwatch/StopwatchViewModel.java
package com.salman.bitclock.ui.stopwatch;

import android.os.Handler;
import android.os.Looper;
import android.os.SystemClock;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.salman.bitclock.data.models.Lap;

import java.util.ArrayList;
import java.util.List;

public class StopwatchViewModel extends ViewModel {

    private MutableLiveData<Long> elapsedTime = new MutableLiveData<>(0L);
    private MutableLiveData<List<Lap>> laps = new MutableLiveData<>(new ArrayList<>());
    private MutableLiveData<Boolean> isRunning = new MutableLiveData<>(false);

    private long startTime = 0L;
    private long timeInMilliseconds = 0L;
    private long timeSwapBuff = 0L;
    private long updateTime = 0L;
    private long lastLapTime = 0L;

    private Handler handler = new Handler(Looper.getMainLooper());

    public LiveData<Long> getElapsedTime() {
        return elapsedTime;
    }

    public LiveData<List<Lap>> getLaps() {
        return laps;
    }

    public LiveData<Boolean> getIsRunning() {
        return isRunning;
    }

    public void start() {
        if (!isRunning.getValue()) {
            startTime = SystemClock.uptimeMillis();
            handler.post(updateTimerThread);
            isRunning.setValue(true);
        }
    }

    public void pause() {
        if (isRunning.getValue()) {
            timeSwapBuff += timeInMilliseconds;
            handler.removeCallbacks(updateTimerThread);
            isRunning.setValue(false);
        }
    }

    public void reset() {
        startTime = 0L;
        timeInMilliseconds = 0L;
        timeSwapBuff = 0L;
        updateTime = 0L;
        lastLapTime = 0L;
        elapsedTime.setValue(0L);
        laps.setValue(new ArrayList<>());
        handler.removeCallbacks(updateTimerThread);
        isRunning.setValue(false);
    }

    public void lap() {
        if (isRunning.getValue()) {
            long currentTime = updateTime;
            long lapTime = currentTime - lastLapTime;
            lastLapTime = currentTime;

            List<Lap> currentLaps = laps.getValue();
            currentLaps.add(0, new Lap(currentLaps.size() + 1, lapTime, currentTime));
            laps.setValue(currentLaps);
        }
    }

    private Runnable updateTimerThread = new Runnable() {
        public void run() {
            timeInMilliseconds = SystemClock.uptimeMillis() - startTime;
            updateTime = timeSwapBuff + timeInMilliseconds;
            elapsedTime.setValue(updateTime);
            handler.postDelayed(this, 10);
        }
    };
}

File: app/src/main/kotlin/com/salman/bitclock/ui/stopwatch/StopwatchViewModel.kt
package com.salman.bitclock.ui.stopwatch

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.salman.bitclock.data.models.Lap
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import javax.inject.Inject

@HiltViewModel
class StopwatchViewModel @Inject constructor() : ViewModel() {

    private val _elapsedTime = MutableStateFlow(0L)
    val elapsedTime = _elapsedTime.asStateFlow()

    private val _isRunning = MutableStateFlow(false)
    val isRunning = _isRunning.asStateFlow()

    private val _laps = MutableStateFlow<List<Lap>>(emptyList())
    val laps = _laps.asStateFlow()

    private var stopwatchJob: Job? = null
    private var lastTickTime = 0L

    fun start() {
        if (_isRunning.value) return
        _isRunning.value = true
        lastTickTime = System.currentTimeMillis()
        stopwatchJob = viewModelScope.launch {
            while (isActive) {
                val now = System.currentTimeMillis()
                _elapsedTime.value += (now - lastTickTime)
                lastTickTime = now
                delay(10)
            }
        }
    }

    fun pause() {
        _isRunning.value = false
        stopwatchJob?.cancel()
    }

    fun lap() {
        val currentElapsed = _elapsedTime.value
        val lapNumber = _laps.value.size + 1
        val lastLapTime = _laps.value.lastOrNull()?.overallTime ?: 0L
        val lapTime = currentElapsed - lastLapTime
        
        val newLap = Lap(lapNumber, lapTime, currentElapsed)
        _laps.value = listOf(newLap) + _laps.value
    }

    fun reset() {
        pause()
        _elapsedTime.value = 0L
        _laps.value = emptyList()
    }

    override fun onCleared() {
        super.onCleared()
        stopwatchJob?.cancel()
    }
}
